src/controller.js

```diff
@@ -1,218 +1,116 @@
-    return msgA != null && msgB != null && msgA.message === msgB.message;
+      .catch(() => this.ovhUserPref.assign('CHATBOT_PREF', this.options));
   }
 
-  static extractDifferenceBetweenArrays(oldArr, newArr) {
-    const last = oldArr.length - 1;
-    for (let i = newArr.length - 1; i >= 0; i -= 1) {
-      if (ChatbotCtrl.areMessagesEquals(newArr[i], oldArr[last])) {
-        // we slice to get the new elements
-        return newArr.slice(i + 1, newArr.length);
-      }
-    }
-    return newArr;
+  pushUserMessage(text) {
+    this.messages.push({ message: text, type: this.MESSAGE_TYPES.user });
   }
 
-  formatMsg(data) {
-    const msg = {
-      message: data.message,
-      items: data.items,
-      buttons: data.buttons.map(button => ({
-        text: button.text,
-        value: button.value,
-        externalUrl: button.type === 'web_url',
-      })),
-    };
-
-    switch (data.origin) {
-      case 'user':
-        msg.type = this.MESSAGE_TYPES.user;
-        break;
-      case 'user_postback':
-        msg.type = this.MESSAGE_TYPES.postback;
-        break;
-      case 'bot':
-        msg.type = this.MESSAGE_TYPES.bot;
-        break;
-      default:
-    }
-    return msg;
+  pushBotMessage(data) {
+    this.messages.push({ message: data.text, type: this.MESSAGE_TYPES.bot });
   }
 
   ask() {
-    const self = this;
     if (!this.message) {
       return;
     }
-    this.$interval.cancel(this.tickInterval);
-    this.tickInterval = undefined;
+
     this.error = null;
-    this.processing = true;
+    this.loaders.asking = true;
+
+    this.pushUserMessage(this.message);
+
     this.chatbotService
       .post(this.message)
       .then((result) => {
-        self.messages = [...self.messages, ...result.data.map(self.formatMsg)];
-        if (!self.tickInterval) {
-          self.tickInterval = self.$interval(self.tick, self.pullRate * 1000);
-        }
-        self.processing = false;
+        this.pushBotMessage(result.data);
+      })
+      .catch(() => {
+        this.error = true;
       })
-      .catch(this.errorHandler);
+      .finally(() => {
+        this.loaders.asking = false;
+      });
 
-    const uMsg = {
-      type: this.MESSAGE_TYPES.user,
-      message: this.message,
-      items: [],
-      buttons: [],
-    };
     this.message = '';
-    this.messages = [...this.messages, uMsg];
-  }
-
-  buttonClick(button) {
-    const self = this;
-    if (button.externalUrl) {
-      this.$window.open(button.value);
-    } else {
-      this.error = null;
-      this.processing = true;
-      this.chatbotService.post(button.value, 'postback')
-        .then((result) => {
-          self.messages = [...self.messages, ...result.data.map(self.formatMsg)];
-          self.processing = false;
-        })
-        .catch(this.errorHandler);
-    }
   }
 
   close() {
     this.hidden = true;
   }
 
   open() {
-    if (!this.initialized) {
-      return this.init();
+    if (!this.started) {
```

Comment:
Looks like this would require a proper state machine :)

Reply:
It looks like we should rather use a state machine to manage state transitions :)